PokeDemoTests/PokemonListPresenterTests.swift

```diff
@@ -0,0 +1,142 @@
+    
+    // MARK: - Tests
+    
+    func testDelegateSetsValidValue(){
+        // Given
+        let mockDelegate = MockPresenterDelegate()
+        
+        // When
+        presenter.setViewDelegate(delegate: mockDelegate)
+        
+        // Then
+        XCTAssertTrue(presenter.delegate is MockPresenterDelegate)
+    }
+    
+    func testGetPokemonListCallsAPI(){
+        // Given
+        let mockService = MockService()
+        
+        // When
+        presenter = PokemonListPresenter(service: mockService)
+        presenter.viewDidLoad()
+        
+        // Then
+        XCTAssertEqual(mockService.sendRequestWithJSONIsCalledIndex, 1)
+        XCTAssertEqual(mockService.sendRequestWithJSONEndPoint, "https://pokeapi.co/api/v2/pokemon")
+    }
+    
+    func testGetPokemonListReturnsSuccessFromAPI(){
+        // Given
+        let mockService = MockService()
+        let mockDelegate = MockPresenterDelegate()
+        presenter = PokemonListPresenter(service: mockService)
+        presenter.setViewDelegate(delegate: mockDelegate)
+        presenter.viewDidLoad()
+        
+        // When service sends successful response
+        let stubPokemons = [Pokemon.init(name: "test", url: "Test URL")]
+//        mockService.onCompletion?(stubPokemons, nil)
+        
+        // Then
+        XCTAssertEqual(mockDelegate.pokemons, stubPokemons)
+        // Check the mockDelegate showPokemonList function is called with right Pokemon array
+    }
+    
+    func testGetPokemonListReturnsErrorFromAPI(){
+        // Given
+        let mockService = MockService()
+        let mockDelegate = MockPresenterDelegate()
+        presenter = PokemonListPresenter(service: mockService)
+        presenter.setViewDelegate(delegate: mockDelegate)
+        presenter.viewDidLoad()
+        
+        // When service sends successful response
+        mockService.onCompletion?(nil, TestError())
+        
+        // Then
+        XCTAssertEqual(mockDelegate.showAlertCalled, true)
+        XCTAssertEqual(mockDelegate.showAlertTitle, Constants.error)
+        XCTAssertEqual(mockDelegate.showAlertMessage, "The operation couldn’t be completed. (PokeDemoTests.TestError error 1.)")
+        // Check the mockDelegate showPokemonList function is called with right Pokemon array
+    }
+    
+    func testGetPokemonListReturnsNoInternetFailureFromAPI(){
+        // Given
+        let mockService = MockService()
+        let mockDelegate = MockPresenterDelegate()
+        presenter = PokemonListPresenter(service: mockService)
+        presenter.setViewDelegate(delegate: mockDelegate)
+        presenter.viewDidLoad()
+        
+        // When service sends successful response
+        mockService.onCompletion?(nil, TestError())
+        
+        // Then
+        XCTAssertEqual(mockDelegate.showAlertCalled, true)
+        XCTAssertEqual(mockDelegate.showAlertTitle, Constants.error)
+        XCTAssertEqual(mockDelegate.showAlertMessage, Internet.shared.noInternet)
+    }
```

Comment:
Isn't it the same test as previous. What's the difference?

Reply:
Covered in 14abf11f45283b232c0d56311c6b93990f9ba842